fontir/src/variations.rs

```diff
@@ -0,0 +1,795 @@
+struct VariationRegion(HashMap<String, Tent>);
+
+impl VariationRegion {
+    fn new() -> Self {
+        VariationRegion(HashMap::new())
+    }
+}
+
+impl VariationRegion {
+    /// The scalar multiplier for the provided location for this region
+    ///
+    /// Returns None for no influence, Some(non-zero value) for influence.
+    ///
+    /// In Python, supportScalar. We only implement the ot=True, extrapolate=False paths.
+    /// <https://github.com/fonttools/fonttools/blob/2f1f5e5e7be331d960a0e30d537c2b4c70d89285/Lib/fontTools/varLib/models.py#L123>.
+    fn scalar_at(&self, location: &NormalizedLocation) -> OrderedFloat<f32> {
+        let scalar =
+            self.0
+                .iter()
+                .filter(|(_, ar)| ar.validate())
+                .fold(ONE, |scalar, (axis_name, tent)| {
+                    let v = location
+                        .get(axis_name)
+                        .map(|v| v.into_inner())
+                        .unwrap_or_default();
+                    let lower = tent.lower.into_inner();
+                    let peak = tent.peak.into_inner();
+                    let upper = tent.upper.into_inner();
+
+                    // If we're at the peak by definition we have full influence
+                    if v == peak {
+                        return scalar; // *= 1
+                    }
+
+                    // If the Tent is 0,0,0 it's always in full effect
+                    if (lower, peak, upper) == (ZERO, ZERO, ZERO) {
+                        return scalar; // *= 1
+                    }
+
+                    if v <= lower || upper <= v {
+                        trace!(
+                            "  {:?} => 0 due to {} {:?} at {:?}",
+                            self,
+                            axis_name,
+                            tent,
+                            location
+                        );
+                        return ZERO;
+                    }
+
+                    let subtract_me = if v < peak {
+                        tent.lower.into_inner()
+                    } else {
+                        tent.upper.into_inner()
+                    };
+                    scalar * (v - subtract_me) / (peak - subtract_me)
+                });
+        scalar
+    }
+}
+
+/// The min/peak/max of a masters influence.
+///
+/// Visualize as a tent of influence, starting at lower, peaking at peak,
+/// and dropping off to zero at upper.
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq)]
+struct Tent {
+    lower: NormalizedCoord,
+    peak: NormalizedCoord,
+    upper: NormalizedCoord,
+}
+
+impl Tent {
+    fn new(mut lower: NormalizedCoord, peak: NormalizedCoord, mut upper: NormalizedCoord) -> Self {
+        let zero = NormalizedCoord::new(0.0);
+        if peak > zero {
+            lower = zero;
+        } else {
+            upper = zero;
+        }
+        Tent { lower, peak, upper }
+    }
+}
+
+impl Tent {
```

Comment:
nit: this `impl Tent {` block can be joined with the one immediately preceding, if I'm not mistaken (which is why the first time I somehow got confused and missed the `Tent::new` implementation)

Reply:
I have a habbit of keeping methods that take self separate of methods that don't but there is no real reason for that. Updated.